flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java

```diff
@@ -100,6 +127,8 @@ protected void log(
         return serializeRaw(
                 wrapper -> {
                     wrapper.writeByte(op.code);
+                    // todo: wrapper.writeShort(stateId); and/or sort and write once (same for key, ns?)
+                    wrapper.writeUTF(metaInfo.getName());
```

Comment:
Writing state name for each change is sub-optimal, and there are two ways to avoid it:
1. write a mapping from state name to some id in the beginning, and write only id here
2. sort changes by state name and write it once

The first option is easier to implement but turned out more error-prone (on recovery).
The second option is also more efficient and can be extended to other dimensions as well (key, namespace).

However, I think it's easier to have this simpler version with recovery integrated (and maybe benchmarked) and then apply the optimization.